contracts/cw721-base/helpers.ts

```diff
@@ -0,0 +1,433 @@
+    const client = await connect(wallet, defaultOptions);
+
+    // ensure we have some tokens
+    if (options.faucetUrl) {
+      const account = await client.getAccount();
+      if (!account) {
+        console.log(`Getting ${options.feeToken} from faucet`);
+        await hitFaucet(options.faucetUrl, client.senderAddress, options.faucetToken);
+      }  
+    }
+
+    return client;
+  }
+
+  const recoverMnemonic = async (password: string, filename?: string): Promise<string> => {
+    const keyfile = filename || options.defaultKeyFile;
+    const wallet = await loadOrCreateWallet(defaultOptions, keyfile, password);
+    return wallet.mnemonic;
+  }
+
+  return {setup, recoverMnemonic};
+}
+
+interface Balances {
+  readonly address: string
+  readonly amount: string  // decimal as string
+}
+
+interface MintInfo {
+  readonly minter: string
+  readonly cap?: string // decimal as string
+}
+
+interface ContractInfo {
+  readonly name: string
+  readonly symbol: string
+}
+
+interface InitMsg {
+  readonly name: string
+  readonly symbol: string
+  readonly minter: string
+}
+// Better to use this interface?
+interface MintMsg {
+  readonly token_id: string
+  readonly owner: string
+  readonly name: string
+  readonly description?: string
+  readonly image?: string
+}
+
+type Expiration = {readonly at_height: number} | {readonly at_time: number} | {readonly never: {}}; 
+
+interface AllowanceResponse {
+  readonly allowance: string;  // integer as string
+  readonly expires: Expiration;
+}
+
+interface AllowanceInfo {
+  readonly allowance: string;  // integer as string
+  readonly spender: string; // bech32 address
+  readonly expires: Expiration;
+}
+
+interface AllAllowancesResponse {
+  readonly allowances: readonly AllowanceInfo[];
+}
+
+interface AllAccountsResponse {
+  // list of bech32 address that have a balance
+  readonly accounts: readonly string[];
+}
+
+interface TokensResponse {
+  readonly tokens: readonly string[];
+}
+
+interface CW721Instance {
+  readonly contractAddress: string
+
+  // queries
+  // balance: (address?: string) => Promise<string>
+  allowance: (owner: string, spender: string) => Promise<AllowanceResponse>
+  allAllowances: (owner: string, startAfter?: string, limit?: number) => Promise<AllAllowancesResponse>
+  allAccounts: (startAfter?: string, limit?: number) => Promise<readonly string[]>
+  contractInfo: () => Promise<any>
+  ownerOf: (owner: string) => Promise<any>
+  nftInfo: (tokenId: string) => Promise<any>
+  allNftInfo: (tokenId: string) => Promise<any>
+  // tokenInfo: () => Promise<any>
+  minter: () => Promise<any>
+  numTokens: () => Promise<any>
+  tokens: (owner:string, start_after: string, limit: number ) => Promise<TokensResponse>
+  allTokens: (start_after?: string, limit?: number ) => Promise<TokensResponse>
+
+  // actions
+  mint: (token_id: string, owner:string, name:string, description?: string) => Promise<string>
```

Comment:
Also takes `image?: string` I believe

Reply:
I think defining `type TokenId = string` somewhere above and using it some places would make the return values clearer - no runtime or type-check difference, just more readable. Actually, what is the string returned here... I guess not the TokenId.

Reply:
Oh, and for public API, nice to use `tokenId` rather than `token_id`